File: application/src/main/java/org/thingsboard/server/controller/WhiteLabelingController.java
/**
 * ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
 *
 * Copyright © 2016-2023 ThingsBoard, Inc. All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of ThingsBoard, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to ThingsBoard, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material is strictly forbidden
 * unless prior written permission is obtained from COMPANY.
 *
 * Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
 * managers or contractors who have executed Confidentiality and Non-disclosure agreements
 * explicitly covering such access.
 *
 * The copyright notice above does not evidence any actual or intended publication
 * or disclosure  of  this source code, which includes
 * information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
 * OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
 * THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
 * AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
 * THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
 * DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
 * OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */
package org.thingsboard.server.controller;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiParam;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.thingsboard.server.common.data.StringUtils;
import org.thingsboard.server.common.data.exception.ThingsboardException;
import org.thingsboard.server.common.data.id.CustomerId;
import org.thingsboard.server.common.data.id.EntityId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.permission.Operation;
import org.thingsboard.server.common.data.permission.Resource;
import org.thingsboard.server.common.data.security.Authority;
import org.thingsboard.server.common.data.wl.LoginWhiteLabelingParams;
import org.thingsboard.server.common.data.wl.WhiteLabelingParams;
import org.thingsboard.server.dao.wl.WhiteLabelingService;
import org.thingsboard.server.queue.util.TbCoreComponent;
import org.thingsboard.server.service.mail.MailTemplates;

import javax.servlet.http.HttpServletRequest;
import java.util.concurrent.ExecutionException;

import static org.thingsboard.server.controller.ControllerConstants.SYSTEM_OR_TENANT_AUTHORITY_PARAGRAPH;
import static org.thingsboard.server.controller.ControllerConstants.CUSTOMER_ID_PARAM_DESCRIPTION;
import static org.thingsboard.server.controller.ControllerConstants.TENANT_AUTHORITY_PARAGRAPH;
import static org.thingsboard.server.controller.ControllerConstants.TENANT_OR_CUSTOMER_AUTHORITY_PARAGRAPH;
import static org.thingsboard.server.controller.ControllerConstants.WL_READ_CHECK;
import static org.thingsboard.server.controller.ControllerConstants.WL_WRITE_CHECK;

@RestController
@TbCoreComponent
@RequestMapping("/api")
public class WhiteLabelingController extends BaseController {

    private static final String LOGO_CHECKSUM_DESC = "Logo image checksum. Expects value from the browser cache to compare it with the value from settings. If value matches, the 'logoImageUrl' will be null.";
    private static final String FAVICON_CHECKSUM_DESC = "Favicon image checksum. Expects value from the browser cache to compare it with the value from settings. If value matches, the 'faviconImageUrl' will be null.";

    @Autowired
    private WhiteLabelingService whiteLabelingService;

    @ApiOperation(value = "Get White Labeling parameters",
            notes = "Returns white-labeling parameters for the current user.", produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/whiteLabelParams", method = RequestMethod.GET, produces = "application/json")
    @ResponseBody
    public WhiteLabelingParams getWhiteLabelParams(
            @ApiParam(value = LOGO_CHECKSUM_DESC)
            @RequestParam(required = false) String logoImageChecksum,
            @ApiParam(value = FAVICON_CHECKSUM_DESC)
            @RequestParam(required = false) String faviconChecksum) throws Exception {
        Authority authority = getCurrentUser().getAuthority();
        WhiteLabelingParams whiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            whiteLabelingParams = whiteLabelingService.getMergedSystemWhiteLabelingParams(TenantId.SYS_TENANT_ID, logoImageChecksum, faviconChecksum);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            whiteLabelingParams = whiteLabelingService.getMergedTenantWhiteLabelingParams(getTenantId(),
                    logoImageChecksum, faviconChecksum);
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            whiteLabelingParams = whiteLabelingService.getMergedCustomerWhiteLabelingParams(getTenantId(),
                    getCurrentUser().getCustomerId(), logoImageChecksum, faviconChecksum);
        }
        return whiteLabelingParams;
    }

    @ApiOperation(value = "Get Login White Labeling parameters",
            notes = "Returns login white-labeling parameters based on the hostname from request.", produces = MediaType.APPLICATION_JSON_VALUE)
    @RequestMapping(value = "/noauth/whiteLabel/loginWhiteLabelParams", method = RequestMethod.GET, produces = "application/json")
    @ResponseBody
    public LoginWhiteLabelingParams getLoginWhiteLabelParams(
            @ApiParam(value = LOGO_CHECKSUM_DESC)
            @RequestParam(required = false) String logoImageChecksum,
            @ApiParam(value = FAVICON_CHECKSUM_DESC)
            @RequestParam(required = false) String faviconChecksum,
            HttpServletRequest request) throws Exception {
        return whiteLabelingService.getMergedLoginWhiteLabelingParams(TenantId.SYS_TENANT_ID, WhiteLabelingService.EDGE_LOGIN_WHITE_LABEL_DOMAIN_NAME, logoImageChecksum, faviconChecksum);
        // TODO: @voba - on edge domain name hardcoded - using login white labeling of the edge owner and not by domain
        // return whiteLabelingService.getMergedLoginWhiteLabelingParams(TenantId.SYS_TENANT_ID, request.getServerName(), logoImageChecksum, faviconChecksum);
    }

    @ApiOperation(value = "Get White Labeling configuration (getCurrentWhiteLabelParams)",
            notes = "Fetch the White Labeling configuration that corresponds to the authority of the user. " +
                    "The API call is designed to load the White Labeling configuration for edition. " +
                    "So, the result is NOT merged with the parent level White Labeling configuration. " +
                    "Let's assume there is a custom White Labeling  configured on a system level. " +
                    "And there is no custom White Labeling  items configured on a tenant level. " +
                    "In such a case, the API call will return default object for the tenant administrator. " +
                    WL_READ_CHECK
            , produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/currentWhiteLabelParams", method = RequestMethod.GET, produces = "application/json")
    @ResponseBody
    public WhiteLabelingParams getCurrentWhiteLabelParams(@ApiParam(value = CUSTOMER_ID_PARAM_DESCRIPTION)
                                                              @RequestParam(value = "customerId", required = false) String strCustomerId) throws ThingsboardException, ExecutionException, InterruptedException {
        Authority authority = getCurrentUser().getAuthority();
        checkWhiteLabelingPermissions(Operation.READ);
        WhiteLabelingParams whiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            whiteLabelingParams = whiteLabelingService.getSystemWhiteLabelingParams(TenantId.SYS_TENANT_ID);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            if (StringUtils.isEmpty(strCustomerId)) {
                whiteLabelingParams = whiteLabelingService.getTenantWhiteLabelingParams(getTenantId());
            } else {
                CustomerId customerId = new CustomerId(toUUID(strCustomerId));
                checkCustomerId(customerId, Operation.READ);
                whiteLabelingParams = whiteLabelingService.getCustomerWhiteLabelingParams(getTenantId(), customerId);
            }
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            whiteLabelingParams = whiteLabelingService.getCustomerWhiteLabelingParams(getTenantId(), getCurrentUser().getCustomerId());
        }
        return whiteLabelingParams;
    }

    @ApiOperation(value = "Get Login White Labeling configuration (getCurrentWhiteLabelParams)",
            notes = "Fetch the Login  White Labeling configuration that corresponds to the authority of the user. " +
                    "The API call is designed to load the Login White Labeling configuration for edition. " +
                    "So, the result is NOT merged with the parent level White Labeling configuration. " +
                    "Let's assume there is a custom White Labeling  configured on a system level. " +
                    "And there is no custom White Labeling  items configured on a tenant level. " +
                    "In such a case, the API call will return default object for the tenant administrator. " +
                    WL_READ_CHECK
            , produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/currentLoginWhiteLabelParams", method = RequestMethod.GET, produces = "application/json")
    @ResponseBody
    public LoginWhiteLabelingParams getCurrentLoginWhiteLabelParams(@ApiParam(value = CUSTOMER_ID_PARAM_DESCRIPTION)
                                                                        @RequestParam(value = "customerId", required = false) String strCustomerId) throws Exception {
        Authority authority = getCurrentUser().getAuthority();
        checkWhiteLabelingPermissions(Operation.READ);
        LoginWhiteLabelingParams loginWhiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            loginWhiteLabelingParams = whiteLabelingService.getSystemLoginWhiteLabelingParams(TenantId.SYS_TENANT_ID);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            if (StringUtils.isEmpty(strCustomerId)) {
                loginWhiteLabelingParams = whiteLabelingService.getTenantLoginWhiteLabelingParams(getTenantId());
            } else {
                CustomerId customerId = new CustomerId(toUUID(strCustomerId));
                checkCustomerId(customerId, Operation.READ);
                loginWhiteLabelingParams = whiteLabelingService.getCustomerLoginWhiteLabelingParams(getTenantId(), customerId);
            }
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            loginWhiteLabelingParams = whiteLabelingService.getCustomerLoginWhiteLabelingParams(getTenantId(), getCurrentUser().getCustomerId());
        }
        return loginWhiteLabelingParams;
    }

    @ApiOperation(value = "Create Or Update White Labeling configuration (saveWhiteLabelParams)",
            notes = "Creates or Updates the White Labeling configuration." +
                    WL_WRITE_CHECK, produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/whiteLabelParams", method = RequestMethod.POST)
    @ResponseStatus(value = HttpStatus.OK)
    public WhiteLabelingParams saveWhiteLabelParams(
            @ApiParam(value = "A JSON value representing the white labeling configuration")
            @RequestBody WhiteLabelingParams whiteLabelingParams,
            @ApiParam(value = CUSTOMER_ID_PARAM_DESCRIPTION)
            @RequestParam(value = "customerId", required = false) String strCustomerId) throws ThingsboardException {
        Authority authority = getCurrentUser().getAuthority();
        checkWhiteLabelingPermissions(Operation.WRITE);
        WhiteLabelingParams savedWhiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            savedWhiteLabelingParams = whiteLabelingService.saveSystemWhiteLabelingParams(whiteLabelingParams);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            if (StringUtils.isEmpty(strCustomerId)) {
                savedWhiteLabelingParams = whiteLabelingService.saveTenantWhiteLabelingParams(getTenantId(), whiteLabelingParams);
            } else {
                CustomerId customerId = new CustomerId(toUUID(strCustomerId));
                checkCustomerId(customerId, Operation.READ);
                savedWhiteLabelingParams = whiteLabelingService.saveCustomerWhiteLabelingParams(getTenantId(), customerId, whiteLabelingParams);
            }
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            savedWhiteLabelingParams = whiteLabelingService.saveCustomerWhiteLabelingParams(getTenantId(), getCurrentUser().getCustomerId(), whiteLabelingParams);
        }
        return savedWhiteLabelingParams;
    }

    @ApiOperation(value = "Create Or Update Login White Labeling configuration (saveWhiteLabelParams)",
            notes = "Creates or Updates the White Labeling configuration." +
                    WL_WRITE_CHECK, produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/loginWhiteLabelParams", method = RequestMethod.POST)
    @ResponseStatus(value = HttpStatus.OK)
    public LoginWhiteLabelingParams saveLoginWhiteLabelParams(
            @ApiParam(value = "A JSON value representing the login white labeling configuration")
            @RequestBody LoginWhiteLabelingParams loginWhiteLabelingParams,
            @ApiParam(value = CUSTOMER_ID_PARAM_DESCRIPTION)
            @RequestParam(name = "customerId", required = false) String strCustomerId) throws Exception {
        Authority authority = getCurrentUser().getAuthority();
        checkWhiteLabelingPermissions(Operation.WRITE);
        LoginWhiteLabelingParams savedLoginWhiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            savedLoginWhiteLabelingParams = whiteLabelingService.saveSystemLoginWhiteLabelingParams(loginWhiteLabelingParams);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            if (StringUtils.isEmpty(strCustomerId)) {
                savedLoginWhiteLabelingParams = whiteLabelingService.saveTenantLoginWhiteLabelingParams(getTenantId(), loginWhiteLabelingParams);
            } else {
                CustomerId customerId = new CustomerId(toUUID(strCustomerId));
                checkCustomerId(customerId, Operation.READ);
                savedLoginWhiteLabelingParams = whiteLabelingService.saveCustomerLoginWhiteLabelingParams(getTenantId(), customerId, loginWhiteLabelingParams);
            }
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            savedLoginWhiteLabelingParams = whiteLabelingService.saveCustomerLoginWhiteLabelingParams(getTenantId(), getCurrentUser().getCustomerId(), loginWhiteLabelingParams);
        }
        return savedLoginWhiteLabelingParams;
    }

    @ApiOperation(value = "Preview Login White Labeling configuration (saveWhiteLabelParams)",
            notes = "Merge the White Labeling configuration with the parent configuration and return the result." +
                    WL_WRITE_CHECK, produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/previewWhiteLabelParams", method = RequestMethod.POST)
    @ResponseStatus(value = HttpStatus.OK)
    public WhiteLabelingParams previewWhiteLabelParams(
            @ApiParam(value = "A JSON value representing the white labeling configuration")
            @RequestBody WhiteLabelingParams whiteLabelingParams) throws ThingsboardException {
        Authority authority = getCurrentUser().getAuthority();
        checkWhiteLabelingPermissions(Operation.WRITE);
        WhiteLabelingParams mergedWhiteLabelingParams = null;
        if (Authority.SYS_ADMIN.equals(authority)) {
            mergedWhiteLabelingParams = whiteLabelingService.mergeSystemWhiteLabelingParams(whiteLabelingParams);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            mergedWhiteLabelingParams = whiteLabelingService.mergeTenantWhiteLabelingParams(getTenantId(), whiteLabelingParams);
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            mergedWhiteLabelingParams = whiteLabelingService.mergeCustomerWhiteLabelingParams(getTenantId(), whiteLabelingParams);
        }
        return mergedWhiteLabelingParams;
    }

    @ApiOperation(value = "Check White Labeling Allowed",
            notes = "Check if the White Labeling is enabled for the current user owner (tenant or customer)" +
                    WL_WRITE_CHECK + TENANT_OR_CUSTOMER_AUTHORITY_PARAGRAPH, produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('TENANT_ADMIN', 'CUSTOMER_USER')")
    @RequestMapping(value = "/whiteLabel/isWhiteLabelingAllowed", method = RequestMethod.GET)
    @ResponseBody
    public Boolean isWhiteLabelingAllowed() throws ThingsboardException {
        Authority authority = getCurrentUser().getAuthority();
        EntityId entityId;
        if (Authority.TENANT_ADMIN.equals(authority)) {
            entityId = getCurrentUser().getTenantId();
        } else {
            entityId = getCurrentUser().getCustomerId();
        }
        return whiteLabelingService.isWhiteLabelingAllowed(getTenantId(), entityId);
    }

    @ApiOperation(value = "Check Customer White Labeling Allowed",
            notes = "Check if the White Labeling is enabled for the customers of the current tenant" +
                    WL_WRITE_CHECK + TENANT_AUTHORITY_PARAGRAPH, produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasAnyAuthority('TENANT_ADMIN')")
    @RequestMapping(value = "/whiteLabel/isCustomerWhiteLabelingAllowed", method = RequestMethod.GET)
    @ResponseBody
    public Boolean isCustomerWhiteLabelingAllowed() throws ThingsboardException {
        return whiteLabelingService.isCustomerWhiteLabelingAllowed(getTenantId());
    }

    @ApiOperation(value = "Save the Mail templates settings (saveMailTemplates)",
            notes = "Creates or Updates the Mail templates settings." + SYSTEM_OR_TENANT_AUTHORITY_PARAGRAPH + WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN')")
    @RequestMapping(value = "/whiteLabel/mailTemplates", method = RequestMethod.POST)
    @ResponseStatus(value = HttpStatus.OK)
    public JsonNode saveMailTemplates(
            @ApiParam(value = "A JSON value representing the Administration Settings.")
            @RequestBody JsonNode mailTemplates) throws Exception {
        checkWhiteLabelingPermissions(Operation.WRITE);
        return whiteLabelingService.saveMailTemplates(getTenantId(), mailTemplates);
    }

    @ApiOperation(value = "Get the Mail templates settings (getMailTemplates)",
            notes = "Fetch Mail template settings. " + SYSTEM_OR_TENANT_AUTHORITY_PARAGRAPH + WL_READ_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN')")
    @RequestMapping(value = "/whiteLabel/mailTemplates", method = RequestMethod.GET)
    @ResponseStatus(value = HttpStatus.OK)
    public JsonNode getMailTemplates(@ApiParam(value = "Use system settings if settings are not defined on tenant level.")
                                         @RequestParam(required = false, defaultValue = "false") boolean systemByDefault) throws Exception {
        checkWhiteLabelingPermissions(Operation.READ);
        JsonNode mailTemplates = whiteLabelingService.getCurrentTenantMailTemplates(getTenantId(), systemByDefault);

        ((ObjectNode) mailTemplates).remove(MailTemplates.API_USAGE_STATE_ENABLED);
        ((ObjectNode) mailTemplates).remove(MailTemplates.API_USAGE_STATE_WARNING);
        ((ObjectNode) mailTemplates).remove(MailTemplates.API_USAGE_STATE_DISABLED);

        return mailTemplates;
    }

    private void checkWhiteLabelingPermissions(Operation operation) throws ThingsboardException {
        accessControlService.checkPermission(getCurrentUser(), Resource.WHITE_LABELING, operation);
    }
}

File: ui/src/app/admin/white-labeling.controller.js
/*
 * ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
 *
 * Copyright © 2016-2018 ThingsBoard, Inc.. All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of ThingsBoard, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to ThingsBoard, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material is strictly forbidden
 * unless prior written permission is obtained from COMPANY.
 *
 * Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
 * managers or contractors who have executed Confidentiality and Non-disclosure agreements
 * explicitly covering such access.
 *
 * The copyright notice above does not evidence any actual or intended publication
 * or disclosure  of  this source code, which includes
 * information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
 * OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
 * THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
 * AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
 * THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
 * DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
 * OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */
import tinycolor from 'tinycolor2';

/* eslint-disable import/no-unresolved, import/default */

import paletteTemplate from './palette-dialog.tpl.html';

/* eslint-enable import/no-unresolved, import/default */

const faviconTypes = ['image/x-icon', 'image/png', 'image/gif', 'image/vnd.microsoft.icon'];
const maxFaviconSize = 262144;
const maxLogoSize = 4194304;

/*@ngInject*/
export default function WhiteLabelingController($state, userService, $scope, $mdDialog, $document, $q,
                                                $translate, toast, whiteLabelingService, $mdTheming, $filter) {
    var vm = this;

    vm.isLoginWl = $state.current.data.isLoginWl;

    vm.maxFaviconSizeKb = maxFaviconSize / 1024;
    vm.maxLogoSizeKb = maxLogoSize / 1024;

    vm.whiteLabelingParams = {};
    vm.primaryPalettes = [];
    vm.accentPalettes = [];

    var palettes = [];
    for (var paletteKey in $mdTheming.PALETTES) {
        if (!paletteKey.startsWith('tb-') && !paletteKey.startsWith('custom-')) {
            var palette = {
                type: paletteKey
            };
            palettes.push(palette);
        }
    }
    vm.primaryPalettes = angular.copy(palettes);
    vm.accentPalettes = angular.copy(palettes);

    vm.paletteName = (palette) => {
        if (palette) {
            return palette.type.toUpperCase().replace('-', ' ');
        } else {
            return "";
        }
    };

    vm.paletteStyle = (palette) => {
        if (palette && palette.type) {
            var key = palette.type == 'custom' ? palette.extends : palette.type;
            var paletteInfo = $mdTheming.PALETTES[key];
            var hex = palette.colors && palette.colors['500']
                    ? palette.colors['500'] : paletteInfo['500'].hex;
            var contrast = paletteInfo['500'].contrast;
            return {
                backgroundColor: hex,
                color: rgba(contrast)
            };
        } else {
            return {};
        }
    };

    vm.paletteTypeChanged = (paletteType, palettes) => {
        var palette = vm.whiteLabelingParams.paletteSettings[paletteType];
        if (palette && palette.type == 'custom') {
            var customPaletteResult = $filter('filter')(palettes, {type: 'custom'}, true);
            if (customPaletteResult && customPaletteResult.length) {
                vm.whiteLabelingParams.paletteSettings[paletteType] = angular.copy(customPaletteResult[0]);
            }
        } else if (palette) {
            delete palette.extends;
            delete palette.colors;
        }
    };

    vm.save = save;
    vm.preview = preview;

    vm.faviconImageAdded = faviconImageAdded;
    vm.clearFaviconImage = clearFaviconImage;
    vm.logoImageAdded = logoImageAdded;
    vm.clearLogoImage = clearLogoImage;
    vm.onFormExit = onFormExit;

    vm.editPrimaryPalette = editPrimaryPalette;
    vm.editAccentPalette = editAccentPalette;

    loadWhiteLabelingParams();

    function loadWhiteLabelingParams() {
        var loadWlPromise = vm.isLoginWl ? whiteLabelingService.getCurrentLoginWhiteLabelParams() : whiteLabelingService.getCurrentWhiteLabelParams();
        loadWlPromise.then(
            (whiteLabelingParams) => {
                vm.whiteLabelingParams = whiteLabelingParams;
                if (!vm.whiteLabelingParams.paletteSettings) {
                    vm.whiteLabelingParams.paletteSettings = {};
                }
                updateCustomPalette('primaryPalette', vm.primaryPalettes);
                updateCustomPalette('accentPalette', vm.accentPalettes);
            }
        );
    }

    function faviconImageAdded($file) {
        var reader = new FileReader();
        reader.onload = function(event) {
            $scope.$apply(function() {
                var dataUrl = event.target.result;
                var type = extractType(dataUrl);
                if (type && faviconTypes.indexOf(type) > -1) {
                    vm.whiteLabelForm.$setDirty();
                    vm.whiteLabelingParams.favicon.url = dataUrl;
                    vm.whiteLabelingParams.favicon.type = extractTypeFromDataUrl(dataUrl);
                } else {
                    toast.showError($translate.instant('white-labeling.favicon-type-error'));
                }
            });
        };
        if ($file.file.size > maxFaviconSize) {
            toast.showError($translate.instant('white-labeling.favicon-size-error', {kbSize: vm.maxFaviconSizeKb}));
            return false;
        } else {
            reader.readAsDataURL($file.file);
        }
    }

    function extractTypeFromDataUrl(dataUrl) {
        var type;
        if (dataUrl) {
            var res = dataUrl.split(";");
            if (res && res.length) {
                res = res[0];
                res = res.split(":");
                if (res && res.length > 1) {
                    type = res[1];
                }
            }
        }
        return type;
    }

    function clearFaviconImage() {
        vm.whiteLabelForm.$setDirty();
        vm.whiteLabelingParams.favicon.url = null;
        vm.whiteLabelingParams.favicon.type = null;
        vm.whiteLabelingParams.faviconChecksum = null;
    }

    function logoImageAdded($file) {
        var reader = new FileReader();
        reader.onload = function(event) {
            $scope.$apply(function() {
                var dataUrl = event.target.result;
                var type = extractType(dataUrl);
                if (type && type.startsWith('image/')) {
                    vm.whiteLabelForm.$setDirty();
                    vm.whiteLabelingParams.logoImageUrl = dataUrl;
                } else {
                    toast.showError($translate.instant('white-labeling.logo-type-error'));
                }
            });
        };
        if ($file.file.size > maxLogoSize) {
            toast.showError($translate.instant('white-labeling.logo-size-error', {kbSize: vm.maxLogoSizeKb}));
            return false;
        } else {
            reader.readAsDataURL($file.file);
        }
    }

    function clearLogoImage() {
        vm.whiteLabelForm.$setDirty();
        vm.whiteLabelingParams.logoImageUrl = null;
        vm.whiteLabelingParams.logoImageChecksum = null;
    }

    function extractType(dataUrl) {
        var type;
        if (dataUrl) {
            var res = dataUrl.split(";");
            if (res && res.length) {
                res = res[0];
                res = res.split(":");
                if (res && res.length > 1) {
                    type = res[1];
                }
            }
        }
        return type;
    }

    function save() {
        var savePromise = vm.isLoginWl ? whiteLabelingService.saveLoginWhiteLabelParams(vm.whiteLabelingParams) :
                        whiteLabelingService.saveWhiteLabelParams(vm.whiteLabelingParams);
        savePromise.then(() => {
            vm.whiteLabelForm.$setPristine();
        });
    }

    function preview() {
        whiteLabelingService.whiteLabelPreview(vm.whiteLabelingParams);
    }

    function onFormExit() {
        whiteLabelingService.cancelWhiteLabelPreview();
    }

    function editPrimaryPalette($event) {
        showEditPaletteDialog($event, vm.whiteLabelingParams.paletteSettings.primaryPalette)
            .then((colors) => {
                updatePaletteColors(vm.whiteLabelingParams.paletteSettings.primaryPalette, colors, vm.primaryPalettes);
            });
    }

    function editAccentPalette($event) {
        showEditPaletteDialog($event, vm.whiteLabelingParams.paletteSettings.accentPalette)
            .then((colors) => {
                updatePaletteColors(vm.whiteLabelingParams.paletteSettings.accentPalette, colors, vm.accentPalettes);
            });
    }

    function updatePaletteColors(palette, colors, palettes) {
        if (colors) {
            palette.colors = colors;
            if (palette.type != 'custom') {
                palette.extends = palette.type;
                palette.type = 'custom';
            }
            var customPaletteResult = $filter('filter')(palettes, {type: 'custom'}, true);
            if (!customPaletteResult || !customPaletteResult.length) {
                palettes.push(angular.copy(palette));
            } else {
                var index = palettes.indexOf(customPaletteResult[0]);
                if (index > -1) {
                    palettes[index] = angular.copy(palette);
                }
            }
        } else {
            delete palette.colors;
            if (palette.type == 'custom') {
                palette.type = palette.extends;
                delete palette.extends;
            }
        }
    }

    function updateCustomPalette(paletteType, palettes) {
        var palette = vm.whiteLabelingParams.paletteSettings[paletteType];
        if (palette && palette.type == 'custom') {
            var customPaletteResult = $filter('filter')(palettes, {type: 'custom'}, true);
            if (!customPaletteResult || !customPaletteResult.length) {
                palettes.push(angular.copy(palette));
            } else {
                var index = palettes.indexOf(customPaletteResult[0]);
                if (index > -1) {
                    palettes[index] = angular.copy(palette);
                }
            }
        }
    }

    function showEditPaletteDialog($event, palette) {
        var deferred = $q.defer();
        if ($event) {
            $event.stopPropagation();
        }
        $mdDialog.show({
            controller: PaletteDialogController,
            controllerAs: 'vm',
            templateUrl: paletteTemplate,
            parent: angular.element($document[0].body),
            locals: {
                palette: angular.copy(palette)
            },
            fullscreen: true,
            skipHide: true,
            targetEvent: $event
        }).then((colors) => {
            if (angular.equals(colors, {})) {
                colors = null;
            }
            vm.whiteLabelForm.$setDirty();
            deferred.resolve(colors);
        }, function () {
            deferred.reject();
        });
        return deferred.promise;
    }

}

function PaletteDialogController($scope, $mdDialog, $mdTheming, $mdColorPicker, $translate, palette) {

    var vm = this;

    vm.palette = palette;
    vm.colors = palette.colors || {};

    var key = palette.type == 'custom' ? palette.extends : palette.type;

    vm.paletteInfo = $mdTheming.PALETTES[key];
    for (var hue in vm.paletteInfo) {
        if (!vm.colors[hue]) {
            var hex = vm.paletteInfo[hue].hex;
            vm.colors[hue] = hex;
        }
    }

    vm.hueStyle = (hue) => {
        if (hue) {
            var hex = vm.colors[hue];
            var contrast = vm.paletteInfo[hue].contrast;
            return {
                backgroundColor: hex,
                color: rgba(contrast)
            };
        } else {
            return {};
        }
    };

    vm.hueName = (hue) => {
        if (hue == '500') {
            return $translate.instant('white-labeling.primary-background');
        }
        if (hue == '600') {
            return $translate.instant('white-labeling.secondary-background');
        }
        if (hue == '300') {
            return $translate.instant('white-labeling.hue1');
        }
        if (hue == '800') {
            return $translate.instant('white-labeling.hue2');
        }
        if (hue == 'A100') {
            return $translate.instant('white-labeling.hue3');
        }
        return '';
    };

    vm.editColor = function ($event, hue) {

        $mdColorPicker.show({
            value: vm.colors[hue],
            defaultValue: vm.colors[hue],
            random: tinycolor.random(),
            clickOutsideToClose: false,
            hasBackdrop: false,
            skipHide: true,
            preserveScope: false,

            mdColorAlphaChannel: false,
            mdColorSpectrum: true,
            mdColorSliders: true,
            mdColorGenericPalette: false,
            mdColorMaterialPalette: true,
            mdColorHistory: false,
            mdColorDefaultTab: 2,

            $event: $event
        }).then(function (color) {
            vm.colors[hue] = tinycolor(color).toHexString();
            $scope.theForm.$setDirty();
        });
    }

    vm.savePalette = savePalette;
    vm.cancel = cancel;

    function normalizeColors(colors) {
        for (var hue in colors) {
            var origHex = vm.paletteInfo[hue].hex;
            if (vm.colors[hue] == origHex) {
                delete vm.colors[hue];
            }
        }
        return colors;
    }

    function cancel() {
        $mdDialog.cancel();
    }

    function savePalette() {
        $scope.theForm.$setPristine();
        normalizeColors();
        $mdDialog.hide(normalizeColors(vm.colors));
    }
}

function rgba(rgbArray) {
    if ( !rgbArray ) return "rgb('0,0,0')";
    var opacity;
    if (rgbArray.length == 4) {
        rgbArray = angular.copy(rgbArray);
        opacity ? rgbArray.pop() : opacity = rgbArray.pop();
    }
    return opacity && (typeof opacity == 'number' || (typeof opacity == 'string' && opacity.length)) ?
    'rgba(' + rgbArray.join(',') + ',' + opacity + ')' :
    'rgb(' + rgbArray.join(',') + ')';
}